nilearn/reporting/_get_clusters_table.py

```diff
@@ -73,40 +73,76 @@ def _identify_subpeaks(data):
+    of mass for those voxels. If the center of mass falls outside the cluster,
+    we instead report the nearest cluster voxel.
     """
-    # Initial identification of subpeaks with minimal minimum distance
     data_max = maximum_filter(data, 3)
     maxima = data == data_max
-    data_min = minimum_filter(data, 3)
-    diff = (data_max - data_min) > 0
-    maxima[diff == 0] = 0
+    zero_mask = data == 0
+    maxima[zero_mask] = 0
+
+    # Don't treat constant patches as maxima unless the entire cluster is
+    # constant (as in a binary cluster).
+    is_constant = np.isclose(data[~zero_mask].max(), data[~zero_mask].min())
+    if not is_constant:
+        data_min = minimum_filter(data, 3)
+        diff = (data_max - data_min) > 0
+        maxima[diff == 0] = 0
 
     labeled, n_subpeaks = label(maxima)
-    labels_index = range(1, n_subpeaks + 1)
+    labels_index = np.arange(1, n_subpeaks + 1)
     ijk = np.array(center_of_mass(data, labeled, labels_index))
     ijk = np.round(ijk).astype(int)
-    vals = np.apply_along_axis(
-        arr=ijk, axis=1, func1d=_get_val, input_arr=data
-    )
     # Determine if all subpeaks are within the cluster
     # They may not be if the cluster is binary and has a shape where the COM is
     # outside the cluster, like a donut.
-    cluster_idx = np.vstack(np.where(labeled)).T.tolist()
-    subpeaks_outside_cluster = [
-        i
-        for i, peak_idx in enumerate(ijk.tolist())
-        if peak_idx not in cluster_idx
-    ]
-    vals[subpeaks_outside_cluster] = np.nan
-    if subpeaks_outside_cluster:
+    subpeaks_outside_cluster = (
+        labeled[ijk[:, 0], ijk[:, 1], ijk[:, 2]] != labels_index
+    )
+    if np.any(subpeaks_outside_cluster):
         warnings.warn(
             "Attention: At least one of the (sub)peaks falls outside of the "
-            "cluster body."
+            "cluster body. Identifying the nearest in-cluster voxel."
         )
+        # Replace centers of mass with their nearest neighbor points in the
+        # corresponding clusters. Note this is also equivalent to computing the
+        # centers of mass constrained to points within the cluster.
+        ijk[subpeaks_outside_cluster] = _cluster_nearest_neighbor(
+            ijk[subpeaks_outside_cluster],
+            labels_index[subpeaks_outside_cluster],
+            labeled
+        )
+    vals = data[ijk[:, 0], ijk[:, 1], ijk[:, 2]]
     return ijk, vals
 
 
+def _cluster_nearest_neighbor(ijk, labels_index, labeled):
+    """Find the nearest neighbor for given points in the corresponding cluster.
+
+    Parameters
+    ----------
+    ijk : :obj:`numpy.ndarray`
+        (n_pts, 3) array of query points.
+    labels_index : `numpy.ndarray`
+        (n_pts,) array of corresponding cluster indices.
+    labeled : `numpy.ndarray`
+        3D array with voxels labeled according to cluster index.
+
+    Returns
+    -------
+    nbrs : `numpy.ndarray`
+        (n_pts, 3) nearest neighbor points.
```

Comment:
Ah, right. Ok I added the tags for the rest of the parameters. I decided not to go through and update all the other docstrings, since that seemed out of scope for this PR. But Lmk if you think it's a good opportunity to do that.

Reply:
Thank you for doing that. As a general rule, we try to limit formatting updates to functions that are directly affected by a given PR, so we don't clutter up the git blame, so I think that was the right choice.